rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -235,6 +235,121 @@ discovery_service:
+matcher includes this, a predefined script for agentless installation will be used for
+the endpoint.
+
+Example agentless config:
+
+```yaml
+discovery_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes" # aws tags to match
+    install:
+      agentless: true
+      # default to this as a result of agentless: true
+      script_name: "default-agentless-installer"
+      sshd_config: "/etc/ssh/sshd_config" # default path
+    ssm:
+      # default to this as a result of agentless: true
+      document_name: "TeleportAgentlessDiscoveryInstaller"
+```
+
+
+An agentless specific SSM document will be required. The `teleport discovery bootstrap`
+command will need to be updated to create SSM documents appropriate for agentless discovery.
+
+Example SSM document:
+```yaml
+# name: TeleportAgentlessDiscoveryInstaller
+---
+schemaVersion: '2.2'
+description: aws:runShellScript
+parameters:
+  caKey:
+    types: String
+    description: "(Required) The Teleport CA to sshd will trust."
+  hostCert:
+    types: String
+    description: "(Required) The Teleport host cert to use."
+  sshdConfigPath:
+    types: String
+    description: "(Required) The path to the sshd config file."
+mainSteps:
+- action: aws:downloadContent
+  name: downloadContent
+  inputs:
+    sourceType: "HTTP"
+    destinationPath: "/tmp/installTeleport.sh"
+    sourceInfo:
+      url: "https://teleportcluster.xyz/webapi/scripts/default-agentless-installer"
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    timeoutSeconds: '300'
+    runCommand:
+      - export CA_KEY='{{ caKey }}'
+      - export HOST_CERT='{{ hostCert }}'
+      - export SSHD_CONFIG='{{ sshdConfigPath }}'
+      - /bin/sh /tmp/installTeleport.sh
+```
+
+Agentless mode will serve a different install script resource named
+`default-agentless-installer`. Which will be used to update and restart the sshd
+configuration.
+
+Possible agentless installer script:
+
+```bash
+(
+  flock -n 9 || exit 1
+
+  if grep -q 'TrustedUserCAKeys /etc/ssh/teleport_user_ca.pub' "$SSHD_CONFIG"; then
+	exit 0
+  fi
+
+  echo "$CA_KEY" > /etc/ssh/teleport_user_ca.pub
+  echo 'TrustedUserCAKeys /etc/ssh/teleport_user_ca.pub' >> $SSHD_CONFIG
+
+  echo 'HostKey /etc/ssh/teleport' >> $SSHD_CONFIG
+  echo 'HostCertificate /etc/ssh/teleport-cert.pub' >> $SSHD_CONFIG
+
+  systemctl restart sshd
+
+) 9>/var/lock/teleport_install.lock
+```
+
+The discovery agent will need CRUD permissions on `host_cert` resources as well as
+permission to create Node resources.
+
+
+### Including AWS Tags as Teleport labels
+
+The AWS tags on discovered EC2 instances will be included as Teleport labels on the
+discovered Nodes.
+
+The labels will be set by the discovery service after the node has successfully
+registered itself. After registration the discovery service will check hourly if the
+available EC2 instances need have their labels updated.
```

Comment:
@lxea After thinking about this some more, I think that this approach we originally discussed is a bit flawed since we want to make the node discovered via API the source of truth and avoid the situation where a node is registered without any labels and only later updated by the discovery service.

To solve this problem I think we might need to introduce a helper resource, e.g. `DiscoveredServer`, which would be used to store metadata about nodes discovered through the API. So the flow would be:

- Discovery service discovers EC2 instances through the API and creates/updates `DiscoveredServer` (name TBD) resources just like other agents heartbeat their resources.
- When Teleport is installed on the instance through SSM, it heartbeats back to the cluster too. When auth server receives the heartbeat from the node, it checks if there's a corresponding `DiscoveredServer` resource (e.g. by matching EC2 instance ID - might need to check if instances report this) and if so, creates a `Server` resource using metadata/labels from `DiscoveredServer` ignoring all labels sent by the node itself in the heartbeat.

What do you think? cc @EdwardDowling